RFM9x.nut

```
@@ -0,0 +1,381 @@
+    
+    
+    constructor(spi, irq, cs=null) {
+        // Assume spi is already initialized, assume cs is already initialized
+        this._spiModule = spi;
+        this._irqPin = irq;
+        this._cs = cs;
+    }
+    
+    function init() {
+        setMode(SLEEP); // need to do this to go into LoRa mode
+        
+        // Defaults
+        setFrequency(RFM9X_DEFAULT_FREQUENCY);
+        setBandwidth(RFM9X_DEFAULT_BANDWIDTH);
+        setSpreadingFactor(RFM9X_DEFAULT_SF);
+        setCodingRate(RFM9X_DEFAULT_CODING_RATE);
+        setImplicitHeaderMode(RFM9X_DEFAULT_IMPLICIT_HEADER_MODE);
+        setPreambleLength(RFM9X_DEFAULT_PREAMBLE_LENGTH);
+
+        _configISR();
+        
+        _setInLoRaMode();
+
+    }
+
+    function _configISR() {
+        if (_irqPin) {
+            _irqPin.configure(DIGITAL_IN_PULLDOWN, _interruptServiceRoutine.bindenv(this));
+        }
+        else {
+            throw "interrupt pin passed to constructor is invalid";
+        }
+    }
+
+    function _interruptServiceRoutine() {
+        // Active high
+        if (_irqPin.read()) {
+            local read = readReg(RFM9X_REG_IRQ_FLAGS);
+            if (read & (1 << PAYLOAD_CRC_ERROR)) {
+                _error = true;
+            }
+            else if (read & (1 << TX_DONE)) {
+                _sendFinished = true;
+            }
+            else if (read == ((1 << RX_DONE) | (1 << VALID_HEADER))) {
+                _receive(_readFromRXBuffer());
+            }
+            
+            clearInterrupts();
+        }
+    }
+
+    function _receive(data) {
+        _receiveHandler && _receiveHandler(data);
+    }
+    
+    function _writeToTXBuffer(data, len) {
+        local write = blob(len + 1);
+        write[0] = (RFM9X_FIFO | RFM9X_WRITE_MASK);
+        for(local i = 0; i < len; ++i) {
+            write[i+1] = data[i];
+        }
+        _csLow();
+        _spiModule.writeread(write);
+        _csHigh();
+    }
+    
+    function sendData(data) {
```

There appear to be many ways this method can fail silently. This is not great for the user.  Need to notify user why a send didn't happen.  

May want to consider adding a callback to this method - a function that is called when data is done sending and that can also have an error parameter if there were any issues found and the message didn't send.